### ISSUE TITLE
IOError: [Errno 2] No such file or directory: '/dev/fd/11'

### ISSUE THREAD

**Issue by ganado** [google/honggfuzz#210]
Hello
while running the fuzzer from the docker with "persistent" mode, I got the following error:

```
root@df0aedc8257e:/# /honggfuzz/honggfuzz -v -c /honggfuzz/examples/externalfuzzers/lowBytesIncrease.py -f seeds/ -- ./run
Persistent signature found in './run'. Enabling persistent fuzzing mode
cmdline: './run' pid: 6463, inputDir 'seeds/', nullifyStdio: true, fuzzStdin: false, saveUnique: true, mutationsPerRun: 6, externalCommand: '/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py', runEndTime: 0 tmOut: 10, mutationsMax: 0, threads.threadsMax: 2, fileExtn: 'fuzz', ASLimit: 0x0(MiB), RSSLimit: 0x0, DATALimit: 0x0, fuzzExe: './run', fuzzedPid: 0, monitorSIGABRT: 'true'
Entering phase 1/2: Dry Run
Launched new fuzzing thread, no. #0
Launched new fuzzing thread, no. #1
Persistent mode: Launched new persistent PID: 6466
Persistent mode: Launched new persistent PID: 6467
[2018-06-15T13:07:21+0000][W][6465] arch_checkWait():309 Persistent mode: PID 6467 exited with status: EXITED, exit code: 1
Size:393478 (i,b,hw,edge,ip,cmp): 0/0/0/25/3/522, Tot:0/0/0/25/3/522
[2018-06-15T13:07:21+0000][W][6464] arch_checkWait():309 Persistent mode: PID 6466 exited with status: EXITED, exit code: 1
Size:393478 (i,b,hw,edge,ip,cmp): 0/0/0/123/8/254, Tot:0/0/0/148/11/776
Persistent mode: Launched new persistent PID: 6468
Persistent mode: Launched new persistent PID: 6469
[2018-06-15T13:07:21+0000][W][6465] arch_checkWait():309 Persistent mode: PID 6468 exited with status: EXITED, exit code: 1
[2018-06-15T13:07:21+0000][W][6464] arch_checkWait():309 Persistent mode: PID 6469 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 6470
Persistent mode: Launched new persistent PID: 6471
[2018-06-15T13:07:21+0000][W][6465] arch_checkWait():309 Persistent mode: PID 6470 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 6472
[2018-06-15T13:07:21+0000][W][6464] arch_checkWait():309 Persistent mode: PID 6471 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 6473
[2018-06-15T13:07:21+0000][W][6465] arch_checkWait():309 Persistent mode: PID 6472 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 6474
[2018-06-15T13:07:21+0000][W][6464] arch_checkWait():309 Persistent mode: PID 6473 exited with status: EXITED, exit code: 1
[2018-06-15T13:07:21+0000][W][6465] arch_checkWait():309 Persistent mode: PID 6474 exited with status: EXITED, exit code: 1
Entering phase 2/2: Dynamic Main
Traceback (most recent call last):
  File "/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py", line 9, in <module>
    with open(sys.argv[1], "r+b") as f:
IOError: [Errno 2] No such file or directory: '/dev/fd/11'
[2018-06-15T13:07:21+0000][E][6464] subproc_System():398 Command '/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py' returned with exit code 1
[2018-06-15T13:07:21+0000][E][6464] input_prepareExternalFile():371 Subprocess '/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py' returned abnormally
[2018-06-15T13:07:21+0000][E][6464] fuzz_fetchInput():382 input_prepareFileExternally() failed
[2018-06-15T13:07:21+0000][F][6464] fuzz_fuzzLoop():437 Cound't prepare input for fuzzing
```

I think the following code seems to cause an error after the file is deleted, so I have a question about this part. How does this code work? I do not understand why this works because it use fd after the file is unlinked. Is this related to error?
```
https://github.com/google/honggfuzz/blob/master/libhfcommon/files.c#L84

int files_writeBufToTmpFile(const char* dir, const uint8_t* buf, size_t fileSz, int flags) {
    char template[PATH_MAX];
    snprintf(template, sizeof(template), "%s/hfuzz.XXXXXX", dir);
    int fd = mkostemp(template, flags);
    if (fd == -1) {
        PLOG_W("mkostemp('%s') failed", template);
        return -1;
    }
    if (unlink(template) == -1) {
        PLOG_W("unlink('%s')", template);
    }
    if (!files_writeToFd(fd, buf, fileSz)) {
        PLOG_W("Couldn't save data to the temporary file");
        close(fd);
        return -1;
    }
    if (lseek(fd, (off_t)0, SEEK_SET) == (off_t)-1) {
        PLOG_W("Couldn't rewind file '%s' fd=%d", template, fd);
        close(fd);
        return -1;
    }
    return fd;
}
```

**Comment by robertswiecki** [comment]
This should be fine, as /
/dev/fd/<x> (which is link to /proc/self/fd/) is a pseudo-symlink. It can point to a deleted file, and it can be opened.

Which operating system are you using? I know that with FreeBSD some kind of /dev/fd filesystem must be mounted first.

**Comment by robertswiecki** [comment]
Ah.. it's docker so Linux. Are you able to log in into your docket container and check the following

```
$ ls -ld /dev/fd
lrwxrwxrwx 1 root root 13 Jun 11 00:02 /dev/fd -> /proc/self/fd
$ ls -ld /proc/self/fd/
dr-x------ 2 jagger jagger 0 Jun 18 00:18 /proc/self/fd/
$ ls -l /proc/self/fd/
total 0
lrwx------ 1 jagger jagger 64 Jun 18 00:18 0 -> /dev/pts/6
lrwx------ 1 jagger jagger 64 Jun 18 00:18 1 -> /dev/pts/6
lrwx------ 1 jagger jagger 64 Jun 18 00:18 2 -> /dev/pts/6
lr-x------ 1 jagger jagger 64 Jun 18 00:18 3 -> /proc/20946/fd
```

**Comment by ganado** [comment]
```
root@cf4183d0db2c:~/test# uname -a
Linux cf4183d0db2c 4.9.87-linuxkit-aufs #1 SMP Wed Mar 14 15:12:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
root@cf4183d0db2c:~/test# cat /etc/issue
Ubuntu 18.04 LTS \n \l

root@cf4183d0db2c:~/test# ls -ld /dev/fd
lrwxrwxrwx 1 root root 13 Jun 18 04:13 /dev/fd -> /proc/self/fd
root@cf4183d0db2c:~/test# ls -ld /proc/self/fd/
dr-x------ 2 root root 0 Jun 18 05:50 /proc/self/fd/
root@cf4183d0db2c:~/test# ls -l /proc/self/fd/
total 0
lrwx------ 1 root root 64 Jun 18 05:50 0 -> /dev/pts/0
lrwx------ 1 root root 64 Jun 18 05:50 1 -> /dev/pts/0
lrwx------ 1 root root 64 Jun 18 05:50 2 -> /dev/pts/0
lr-x------ 1 root root 64 Jun 18 05:50 3 -> /proc/820/fd
root@cf4183d0db2c:~/test# 
```
The existing error seems to have occurred when fuzzer was run on the docker's mounted host volume. When I move the directory into the container and run it, I get another error like this:

```
root@cf4183d0db2c:~/test# /root/honggfuzz/honggfuzz -v -c /root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py -f seeds/ -- ./run
Persistent signature found in './run'. Enabling persistent fuzzing mode
cmdline: './run' pid: 790, inputDir 'seeds/', nullifyStdio: true, fuzzStdin: false, saveUnique: true, mutationsPerRun: 6, externalCommand: '/root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py', runEndTime: 0 tmOut: 10, mutationsMax: 0, threads.threadsMax: 2, fileExtn: 'fuzz', ASLimit: 0x0(MiB), RSSLimit: 0x0, DATALimit: 0x0, fuzzExe: './run', fuzzedPid: 0, monitorSIGABRT: 'true'
Entering phase 1/2: Dry Run
Launched new fuzzing thread, no. #0
Launched new fuzzing thread, no. #1
Persistent mode: Launched new persistent PID: 793
Persistent mode: Launched new persistent PID: 794
[2018-06-18T05:43:54+0000][W][791] arch_checkWait():309 Persistent mode: PID 793 exited with status: EXITED, exit code: 1
Size:1 (i,b,hw,edge,ip,cmp): 0/0/0/148/11/744, Tot:0/0/0/148/11/744
Persistent mode: Launched new persistent PID: 795
[2018-06-18T05:43:54+0000][W][792] arch_checkWait():309 Persistent mode: PID 794 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 796
[2018-06-18T05:43:54+0000][W][791] arch_checkWait():309 Persistent mode: PID 795 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 797
[2018-06-18T05:43:54+0000][W][792] arch_checkWait():309 Persistent mode: PID 796 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 798
[2018-06-18T05:43:54+0000][W][791] arch_checkWait():309 Persistent mode: PID 797 exited with status: EXITED, exit code: 1
Persistent mode: Launched new persistent PID: 799
[2018-06-18T05:43:54+0000][W][792] arch_checkWait():309 Persistent mode: PID 798 exited with status: EXITED, exit code: 1
[2018-06-18T05:43:54+0000][W][791] arch_checkWait():309 Persistent mode: PID 799 exited with status: EXITED, exit code: 1
Entering phase 2/2: Dynamic Main
Traceback (most recent call last):
  File "/root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py", line 10, in <module>
    map = mmap.mmap(f.fileno(), 0)
ValueError: cannot mmap an empty file
Traceback (most recent call last):
  File "/root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py", line 10, in <module>
    map = mmap.mmap(f.fileno(), 0)
ValueError: cannot mmap an empty file
[2018-06-18T05:43:54+0000][E][791] subproc_System():398 Command '/root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py' returned with exit code 1
[2018-06-18T05:43:54+0000][E][791] input_prepareExternalFile():371 Subprocess '/root/honggfuzz/examples/externalfuzzers/lowBytesIncrease.py' returned abnormally
[2018-06-18T05:43:54+0000][E][791] fuzz_fetchInput():382 input_prepareFileExternally() failed
[2018-06-18T05:43:54+0000][F][791] fuzz_fuzzLoop():437 Cound't prepare input for fuzzing
```

**Comment by ganado** [comment]
[strace.log](https://github.com/google/honggfuzz/files/2110041/strace.log)

**Comment by robertswiecki** [comment]
ok.. so it seems that your code tries (actually refueses because of a check) to mmap() an empty file (which, indeed, can be created by honggfuzz). I guess it's something for your code (lowBytesIncrease.py) to handle.

if (size == 0) {
 return early;
}

or somesuch. Though, if I remember correctly mmap()ing of 0-sized files is legal.

In any case, the tool (lowBytesIncrease.py) should expect empty files (especially with -c, where it's up to the tool to create contents of the file).

**Comment by robertswiecki** [comment]
Also, it seems that you use -c which basically delivers en empty file to the command, maybe you want to use --pprocess_cmd which operates on mutated files? In any case, your tool should also expect 0-sized files with --pprocess_cmd sometimes.

**Comment by will** [comment]
I'm trying to run the badcode external fuzzer example and am running into to this now. Ubuntu, no docker. It seems like the included python script in examples/externalfuzzers needs to change to handle empty files? I'm new to all of this so still poking around

**Comment by robertswiecki** [comment]
Yeah.. I think so, are you using this script for something, or just poking around using it. My point is that it was written by one of the honggfuzz contributors early on, and I'm not sure how useful it is currently?

**Comment by will** [comment]
I wanted to get an example working that found a crash to see if I could get anything to work as a first step. I saw the "badcode" directory and figured that'd be the best place to start.

I tried adding 
```python
if os.stat(sys.argv[1]).st_size == 0:
    exit()
else:
    print("whoo not exiting!")
```

before the `with open`… but it appears it never gets a non-empty file. At least for the few minutes I left it running.

Instead I made the buffer in badcode even smaller and that got it to trigger finding a crash, so I think I'll leave this python thing alone.
